### ISSUE TITLE
Expose display name in TestListener API

### ISSUE THREAD

**Issue by radarsh** [gradle/gradle#5117]
### Expected Behavior

JUnit 5 provides an annotation, [`@DisplayName`](https://junit.org/junit5/docs/5.0.3/api/org/junit/jupiter/api/DisplayName.html) to describe a test / suite in a human friendly manner. Up until Gradle 4.6, this annotation was honoured by Gradle and its value was being presented in `TestDescriptor.getName()`.

### Current Behavior

Gradle 4.7 seems to have broken this functionality and now it simply gives us the actual name of the method instead of its display name.

### Context

I am the author of a [test logger plugin](https://github.com/radarsh/gradle-test-logger-plugin) and as part of testing it, I run all types of tests (Spock, JUnit 4, JUnit 5 Vintage and JUnit 5 Jupiter tests) and verify the rendered output. This is how I found out about this problem.

### Steps to Reproduce (for bugs)

Have a look at the master branch of my [demo project](https://github.com/radarsh/gradle-junit5-bug). This branch is using Gradle 4.7. If you clone it and run `./gradlew test`, the output we get is

`Test thisTestShouldPass()`

Now, if we switch to the branch [`gradle-4.6`](https://github.com/radarsh/gradle-junit5-bug/tree/gradle-4.6) and run the same task, we get

`Test this test should pass`

Gradle 4.6 behaviour is the correct expected behaviour.

![image](https://user-images.githubusercontent.com/534512/39053759-b281c90e-44a7-11e8-9bd3-760f321feb46.png)

**Comment by marcphilipp** [comment]
Gradle 4.7 now actually tracks display name and method name separately and outputs it into two different columns in the HTML report.

@blindpirate Do you recall why the console output was changed to the method name? For consistency with the other frameworks?

**Comment by blindpirate** [comment]
@radarsh @marcphilipp Thanks.

The reason of this inconsistent behavior is [this line](https://github.com/radarsh/gradle-junit5-bug/blob/master/build.gradle#L17):

```
    afterTest { descriptor ->
        logger.lifecycle "Test ${descriptor.name}"
    }
```

In 4.7 we decided to clarify the concept of `name` and `display name`, and added `getDisplayName()` to `TestDescriptorInternal`. In this case, `descriptor.name` is `thisTestShouldPass` and `descriptor.displayName` is `this test should pass`. We're planning to make `displayName` a public API in a few months so I think it's fine to use it in gradle scripts (Groovy scripts).

Let's leave this issue open in order to track it.

**Comment by radarsh** [comment]
@marcphilipp it's probably not changing the console output directly but the result of `TestDescriptor.getName()` is what's changed between the versions. Now, it depends on whether Gradle / plugins use this method to render on the console or not. 

Either way, I think it warrants another method on the `TestDescriptor` called, say, `getDisplayName()` that gives any human readable values set by test frameworks. The `getName()` could continue to print the actual method name. This would also address another issue I noticed which is that Gradle ignores the `@DisplayName` annotation if set on test classes.

**Comment by radarsh** [comment]
@blindpirate so why isn't it exposed in the public `TestDescriptor` API?

**Comment by radarsh** [comment]
Thanks for clarifying @marcphilipp @blindpirate. I'll wait for the public API to become available.

**Comment by blindpirate** [comment]
@radarsh Because we're not prepared - we definitely want to redesign the test infrastructure but not finished yet. There's an epic issue to track this: https://github.com/gradle/gradle/issues/828

**Comment by tobias-jenkner** [comment]
Can you output the displayname aswell in the xml-reports as attribute of the testcase element:
`      <testcase name="test" classname="org.sample.TestCase" time="0.0" dynamicTest="true" displayname="Sample Test 0 - foo bar" uniqueid="[engine:junit-jupiter]/[class:org.sample.TestCase]/[test-factory:test()]/[dynamic-test:#1]"/>
`
(this Sample in an excerpt of an export from the Eclipse JUnit view)

**Comment by StefMa** [comment]
@blindpirate isn't it enough to add `public String getDisplayName()` to the `TestDescriptor` interface to receive the make it public? 🤔

**Comment by marcphilipp** [comment]
This was resolved in #11026.
